x/archivedb/key.go

```diff
@@ -12,77 +13,135 @@ import (
+//
+// By definition the values associated with a dbKey are append only, there is no
+// support to delete or update it. A new new record with the inner key can be
+// stored, but that will have a different dbKey because the heights will be
+// different. The entries defined under dbMetaKey can be updated onsight.
+//
+// There is no external API to use the dbMetaKey, for now it is reserved for
+// internal usage only.
+type databaseKey interface {
+	InnerKey() []byte
+	Bytes() []byte
+	Height() uint64
+}
+
+// dbKey
 //
 // The keys contains a few extra information alongside the given key. The key is
 // what is known outside of this internal scope as the key, but this struct has
-// more information compacted inside the keyInternal to take advantage of natural
+// more information compacted inside the dbKey to take advantage of natural
 // sorting. The inversed height is stored right after the Prefix, which is
-// MaxUint64 minus the desired Height. An extra byte is also packed inside the
-// key to indicate if the key is a deletion of any previous value (because
-// archivedb is an append only database).
+// MaxUint64 minus the desired Height.
 //
-// Any other property are to not serialized but they are useful when parsing a
-// keyInternal struct from the database
-type keyInternal struct {
-	key       []byte
-	height    uint64
-	isDeleted bool
+// All keys are prefixed with a VarUint64, which is the length of the keys, the
+// idea is to make reads even simpler, making it effectively a O(1) operation.
+// Before this change it would have been O(1) unless a another key existed with
+// the same prefix as the requested. By appending the length of the inner key as
+// the first bytes, it is not longer possible to query by a prefix (the full key
+// is needed to do a lookup) but all reads are a O(1).
+type dbKey struct {
+	key    []byte
+	height uint64
 }
 
 // Creates a new Key struct with a given key and its height
-func newInternalKey(key []byte, height uint64) *keyInternal {
-	return &keyInternal{
-		key:       key,
-		isDeleted: false,
-		height:    height,
+func newKey(key []byte, height uint64) *dbKey {
+	return &dbKey{
+		key:    key,
+		height: height,
 	}
 }
 
-func (k *keyInternal) Bytes() []byte {
+func (k *dbKey) Bytes() []byte {
 	keyLen := len(k.key)
-	bytes := make([]byte, keyLen+internalKeySuffixLen+internalKeyPrefixLen)
-	copy(bytes[0:], internalKeyPrefix)
-	copy(bytes[internalKeyPrefixLen:], k.key)
-	binary.BigEndian.PutUint64(bytes[internalKeyPrefixLen+keyLen:], math.MaxUint64-k.height)
-	if k.isDeleted {
-		bytes[internalKeyPrefixLen+keyLen+longLen] = 1
-	} else {
-		bytes[internalKeyPrefixLen+keyLen+longLen] = 0
-	}
+	keyLenBytes := binary.AppendUvarint([]byte{}, uint64(keyLen))
+
+	bytes := make([]byte, len(keyLenBytes)+keyLen+internalKeySuffixLen)
+	offset := copy(bytes[0:], keyLenBytes)
+	offset += copy(bytes[offset:], k.key)
+	binary.BigEndian.PutUint64(bytes[offset:], math.MaxUint64-k.height)
```

Comment:
Optional - but seems simpler to me
```suggestion
	binary.BigEndian.PutUint64(bytes[offset:], ^k.height)
```